saleor/dashboard/category/views.py

```diff
@@ -0,0 +1,73 @@
+from django.contrib import messages
+from django.core.urlresolvers import reverse_lazy
+from django.shortcuts import get_object_or_404, redirect
+from django.template.response import TemplateResponse
+from django.utils.translation import ugettext_lazy as _
+from django.views.generic import DeleteView
+
+from ...product.models import Category
+from ..views import StaffMemberOnlyMixin, staff_member_required
+from .forms import CategoryForm
+
+
+@staff_member_required
+def category_list(request, root=None):
+    ctx = {}
+    categories = Category.objects.all()
+    if root:
+        current_node = get_object_or_404(Category, pk=root)
+        categories = current_node.get_descendants()
+        ctx['current_node'] = current_node
+        ctx['category_breadcrumbs'] = current_node.get_ancestors(include_self=True)
+    min_level = categories[0].get_level() if categories else 0
```

Comment:
What is this sorcery? `… if categories …` evaluates the entire queryset. That is not the best idea.

![Definitely not the best](http://38.media.tumblr.com/tumblr_md38vseBwE1rk9vano1_r1_500.gif)